### PR TITLE
Feat: 256+ Merkle trie with proofs of inclusion 

### PR DESCRIPTION
**Pull request by lunfardo314**
# Description of change

- implemented `trie` package, a generic trie with the branching factor of 256 and variable length keys
- implemented `trie_merkle` package, a Merkle trie with hash commitments
- implemented Proofs Of Inclusion (and absence) of arbitrary key/value pairs in the chain's state based on Merkle proofs
- minor adjustments and optimizations in `state` and `kv` package

## Links to any relevant issues

* [Commitments to the state](https://hackmd.io/@Evaldas/ryQFvhXfF)
* [256+ trie definition](https://hackmd.io/@Evaldas/H13YFOVGt)

## Type of change
- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

Covered with multiple tests and benchmarks

### CODE REVIEW

File: packages/kv/collections/array32.go [hunk 1]
```diff
@@ -35,16 +35,16 @@ func NewArray32ReadOnly(kvReader kv.KVStoreReader, name string) *ImmutableArray3
 
 const array32ElemKeyCode = byte('#')
 
-func array32SizeKey(name string) kv.Key {
-	return kv.Key(name)
+func Array32SizeKey(name string) []byte {
+	return []byte(name)
```
Comment (dessaya) [inline packages/kv/collections/array32.go]: Why change to `[]byte`? Same for `Array32ElemKey`, `MapSizeKey`, etc. If you need `[]byte` you can always convert after calling, but most callers will expect `kv.Key`.

File: packages/kv/diff.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
+package kv
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+)
+
+// used mostly for testing
```
Comment (dessaya) [inline packages/kv/diff.go:9]: Is this being used? I don't find any callers to functions in this file.

If it's used only for testing it should probably go outside the `kv` package, maybe in something like `kvtest` or `kvdiff` or `kvstats`.

File: packages/kv/kv.go [hunk 3]
```diff
@@ -3,12 +3,13 @@ package kv
...
 	"fmt"
-
 	"github.com/iotaledger/wasp/packages/util"
 	"golang.org/x/xerrors"
+	"io"
+	"os"
 )
 
-// Since map cannot have []byte as key, to avoid unnecessary conversions
+// Key since map cannot have []byte as key, to avoid unnecessary conversions
```
Comment (dessaya) [inline packages/kv/kv.go]: Is this change intentional? The sentence now does not make sense :stuck_out_tongue: 

```suggestion
// Key represents a key in the KVStore.
// Since map cannot have []byte as key, to avoid unnecessary conversions
```

File: packages/kv/kv.go [hunk 4]
```diff
@@ -155,3 +156,65 @@ func ByteSize(s KVStoreReader) int {
...
+
+func DumpToFile(r KVStoreReader, fname string) (int, error) {
+	file, err := os.Create(fname)
+	if err != nil {
+		return 0, err
+	}
+	defer file.Close()
+
+	var bytesTotal int
+	_ = r.Iterate("", func(k Key, v []byte) bool {
```
Comment (dessaya) [inline packages/kv/kv.go]: Either capture the error or use `MustIterate`?

File: packages/kv/trie/node.go [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+		n.Terminal = nil
+	}
+	if smallFlags&hasChildrenFlag != 0 {
+		var flags [32]byte
+		if _, err := r.Read(flags[:]); err != nil {
+			return err
+		}
+		for i := 0; i < 256; i++ {
+			ib := uint8(i)
+			if flags[i/8]&(0x1<<(i%8)) != 0 {
```
Comment (dessaya) [inline packages/kv/trie/node.go]: :exploding_head:

File: packages/kv/trie/nodestore.go [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+	}
+	return tr.Model().CommitToNode(n)
+}
+
+func EqualCommitments(c1, c2 CommitmentBase) bool {
+	if c1 == c2 {
+		return true
+	}
+	c1Nil := c1 == nil || (reflect.ValueOf(c1).Kind() == reflect.Ptr && reflect.ValueOf(c1).IsNil())
+	c2Nil := c2 == nil || (reflect.ValueOf(c2).Kind() == reflect.Ptr && reflect.ValueOf(c2).IsNil())
```
Comment (dessaya) [inline packages/kv/trie/nodestore.go:70]: Why is this needed?

File: packages/state/block_test.go [hunk 7]
```diff
@@ -63,28 +65,26 @@ func TestBatches(t *testing.T) {
...
-
-	require.True(t, b.Timestamp().IsZero())
-	require.True(t, b1.Timestamp().IsZero())
-	require.True(t, b2.Timestamp().IsZero())
-
-	require.EqualValues(t, hashing.NilHash, b.PreviousStateHash())
-	require.EqualValues(t, hashing.NilHash, b1.PreviousStateHash())
-	require.EqualValues(t, hashing.NilHash, b2.PreviousStateHash())
-}
+//func TestOriginBlock(t *testing.T) {
```
Comment (dessaya) [inline packages/state/block_test.go]: Why is this test commented out?

Suggest using `t.Skip("skipping because this and that")` instead of commenting out. (Or if it doesn't compile just comment out the offending block and skip), so that when running tests we can know how many tests are actually skipped.

File: packages/state/types.go [hunk 8]
```diff
@@ -14,13 +14,16 @@ import (
 type VirtualStateAccess interface {
 	BlockIndex() uint32
 	Timestamp() time.Time
-	PreviousStateHash() hashing.HashValue
-	StateCommitment() hashing.HashValue
+	TrieAccess() trie.NodeStore
```
Comment (dessaya) [inline packages/state/types.go:17]: I don't understand what "Access" means in "VirtualStateAccess", "TrieAccess", etc. Why not call this just `Trie()`?

**Review by dessaya — APPROVED** [review]

File: packages/kv/diff.go (hunk 2 above)
Comment (lunfardo314) [reply packages/kv/diff.go:9]: will delete a bit later. It is need for deep debugging, may be usefull in the future

File: packages/kv/collections/array32.go (hunk 1 above)
Comment (lunfardo314) [reply packages/kv/collections/array32.go]: this is a confusion which one to use where and what is the main use case. Just made it uniform after exposed it as an exported function (before it was internal). Intended use case is composing of keys in proofs of inclusion, we need bytes there

File: packages/kv/kv.go (hunk 3 above)
Comment (lunfardo314) [reply packages/kv/kv.go]: don't remember, what's wrong here?

File: packages/kv/collections/array32.go (hunk 1 above)
Comment (dessaya) [reply packages/kv/collections/array32.go]: I still think that for consistency all functions that deal with keys must use `kv.Key`. If you need bytes, you can do `[]byte(Array32SizeKey("..."))`

File: packages/kv/kv.go (hunk 3 above)
Comment (dessaya) [reply packages/kv/kv.go]: "Key since map cannot have []byte as key, [...]"
The problem is that the sentence does not make sense like this. Godoc comments must be simple English sentences. Starting with the same identifier is a recommendation, but not a requirement. Especially not if it breaks the sentence.
